File: 12장 그래프/src/main/java/com/algo/graph/leetcode_17/LeetCode_17.java
package com.algo.graph.leetcode_17;

import lombok.RequiredArgsConstructor;

import java.util.List;
import java.util.Map;

@RequiredArgsConstructor
public class LeetCode_17 {
    private final Map<Character, List<Character>> map;

    public void solution(String input, List<String> result, int index, StringBuilder path) {
        if (path.length() == input.length()) {
            result.add(path.toString());
            return;
        }

        for (Character c : map.get(input.charAt(index)))
            solution(input, result, index + 1, new StringBuilder(path).append(c));
    }
}

File: 12장 그래프/src/main/java/com/algo/graph/LeetCode17.kt
package com.algo.graph

class LeetCode17(private val map: Map<Char, List<Char>>) {
    fun solution(input: String): List<String> {
        val result: MutableList<String> = mutableListOf()

        fun dfs(index: Int, path: StringBuilder) {
            if (index == input.length) {
                result.add(path.toString())
                return
            }

            for (c in map[input[index]]!!)
                dfs(index + 1, StringBuilder(path).append(c))
        }

        dfs(0, StringBuilder())

        return result
    }
}


